pkg/localstorev2/internal/stampindex/stamp.go

```diff
@@ -0,0 +1,163 @@
+
+// stampIndexItem is an store.Item that represents data relevant to stamp.
+type stampIndexItem struct { // stampIndexItem
+	namespace      []byte // The namespace of other related item.
+	batchID        []byte
+	batchIndex     []byte
+	batchTimestamp []byte
+	chunkImmutable bool
+}
+
+// ID implements the storage.Item interface.
+func (i stampIndexItem) ID() string {
+	return fmt.Sprintf("%s/%s/%s", string(i.namespace), string(i.batchID), string(i.batchIndex))
+}
+
+// Namespace implements the storage.Item interface.
+func (i stampIndexItem) Namespace() string {
+	return "stampIndex"
+}
+
+// Marshal implements the storage.Item interface.
+func (i stampIndexItem) Marshal() ([]byte, error) {
+	switch {
+	case len(i.batchID) != swarm.HashSize:
+		return nil, errStampItemMarshalBatchIDInvalid
+	case len(i.batchIndex) != swarm.StampIndexSize:
+		return nil, errStampItemMarshalBatchIndexInvalid
+	case len(i.namespace) == 0:
+		return nil, errStampItemMarshalNamespaceInvalid
+	}
+
+	var chImmutable byte = '0'
+	if i.chunkImmutable {
+		chImmutable = '1'
+	}
+
+	nsLen := len(i.namespace)
+	buf := make([]byte, 8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1)
+
+	binary.LittleEndian.PutUint64(buf, uint64(len(i.namespace)))
+	copy(buf[8:], i.namespace)
+	copy(buf[8+nsLen:], i.batchID)
+	copy(buf[8+nsLen+swarm.HashSize:], i.batchIndex)
+	copy(buf[8+nsLen+swarm.HashSize+swarm.StampIndexSize:], i.batchTimestamp)
+	buf[8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1] = chImmutable
+	return buf, nil
+}
+
+// Unmarshal implements the storage.Item interface.
+func (i *stampIndexItem) Unmarshal(bytes []byte) error {
+	if len(bytes) < 8 {
+		return errStampItemUnmarshalInvalidSize
+	}
+	nsLen := int(binary.LittleEndian.Uint64(bytes))
+	if len(bytes) != 8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1 {
+		return errStampItemUnmarshalInvalidSize
+	}
+
+	ni := new(stampIndexItem)
+	ni.namespace = append(make([]byte, 0, nsLen), bytes[8:nsLen]...)
+	ni.batchID = append(make([]byte, 0, swarm.HashSize), bytes[8+nsLen:8+nsLen+swarm.HashSize]...)
+	ni.batchIndex = append(make([]byte, 0, swarm.StampIndexSize), bytes[8+nsLen+swarm.HashSize:8+nsLen+swarm.HashSize+swarm.StampIndexSize]...)
+	ni.batchTimestamp = append(make([]byte, 0, swarm.StampTimestampSize), bytes[8+nsLen+swarm.HashSize+swarm.StampIndexSize:8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize]...)
+	switch bytes[8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize] {
+	case '0':
+		ni.chunkImmutable = false
+	case '1':
+		ni.chunkImmutable = true
+	default:
+		return errStampItemUnmarshalChunkImmutableInvalid
+	}
+	*i = *ni
+	return nil
+}
+
+// String implements the fmt.Stringer interface.
+func (i stampIndexItem) String() string {
+	return path.Join(i.Namespace(), i.ID())
+}
+
+func LoadOrStore(s internal.Storage, namespace string, chunk swarm.Chunk) ([]byte, bool, error) {
```

Comment:
This call should also return a boolean which says if the value was loaded or not.

If the value was loaded we need to check the return values. If not, we can go ahead with putting the chunk.

Reply:
Done.